File: go/tournament/tournament.go
package tournament

import (
	"bufio"
	"fmt"
	"io"
	"sort"
	"strings"
)

type standings struct {
	team   string
	played int
	won    int
	draw   int
	lost   int
	points int
}

// Tally tallies the results of a football tournament
func Tally(r io.Reader, w io.Writer) error {
	var err error
	var tallyMap = make(map[string]standings)

	scanner := bufio.NewScanner(r)
	// Read each line one at a time
	for scanner.Scan() {
		game := scanner.Text()
		// Ignore empty or commented lines
		if len(game) == 0 || strings.HasPrefix(game, "#") {
			continue
		}
		g := strings.Split(game, ";")
		if len(g) != 3 {
			return fmt.Errorf("input should be in the form team1;team2;outcome. got: %s", game)
		}
		team1, team2, outcome := g[0], g[1], g[2]
		a := tallyMap[team1]
		b := tallyMap[team2]

		a.team = team1
		b.team = team2
		// Matches played
		a.played++
		b.played++

		switch outcome {
		case "win":
			a.points += 3
			a.won++
			b.lost++
		case "loss":
			b.points += 3
			b.won++
			a.lost++
		case "draw":
			a.points++
			a.draw++
			b.points++
			b.draw++
		default:
			return fmt.Errorf("outcome should be one of win,loss,draw. got: %s", outcome)
		}
		tallyMap[team1] = a
		tallyMap[team2] = b
	}

	// Extract the standings into a slice
	tally := make([]standings, len(tallyMap))
	i := 0
	for _, v := range tallyMap {
		tally[i] = v
		i++
	}
	// Sort
	sort.Slice(tally, func(i, j int) bool {
		if tally[i].points == tally[j].points {
			return tally[i].team < tally[j].team
		}
		return tally[i].points > tally[j].points
	})
	// Format string and send it to writer
	_, err = fmt.Fprint(w, "Team                           | MP |  W |  D |  L |  P\n")
	if err != nil {
		return err
	}
	var table = "|  %d |  %d |  %d |  %d |  %d\n"
	for _, res := range tally {
		_, err = fmt.Fprintf(
			w, res.team+strings.Repeat(" ", 31-len(res.team))+table,
			res.played, res.won, res.draw, res.lost, res.points,
		)
		if err != nil {
			return err
		}
	}
	return nil
}

File: python/tournament/tournament.py
from collections import defaultdict


def tally(rows: list) -> list:
    # transform the row data into actual results
    results = defaultdict(lambda: defaultdict(int))
    for row in rows:
        team_a, team_b, outcome = row.split(";")

        results[team_a]["matches_played"] += 1
        results[team_b]["matches_played"] += 1

        if outcome == "win":
            results[team_a]["wins"] += 1
            results[team_b]["losses"] += 1
            results[team_a]["points"] += 3
        if outcome == "loss":
            results[team_b]["wins"] += 1
            results[team_a]["losses"] += 1
            results[team_b]["points"] += 3
        if outcome == "draw":
            results[team_a]["draws"] += 1
            results[team_a]["points"] += 1
            results[team_b]["draws"] += 1
            results[team_b]["points"] += 1

    # sort the results by points(descending) and alphabetically
    sorted_results = sorted(results, key=lambda team: (-results[team]["points"], team))

    # use the sorted results to generate the final output
    tally_table = ["{:<31}".format("Team") + "| MP |  W |  D |  L |  P"]
    for team in sorted_results:
        points = "|  {MP} |  {W} |  {D} |  {L} |  {P}".format(
            MP=results[team]["matches_played"],
            W=results[team]["wins"],
            D=results[team]["draws"],
            L=results[team]["losses"],
            P=results[team]["points"],
        )
        tally_table.append("{:<31}".format(team) + points)

    return tally_table
